### DISCUSSION TITLE
How to unify fonts and color schemes across systems in the context of snapshot testing?

### DISCUSSION THREAD

**Discussion by jwedel** [simonsmith/cypress-image-snapshot#39] [General]
This is not directly a "bug report" but rather a question for best practices. Maybe the answer could be added to the README.

What we've seen is, that between Windows, Mac and Linux, there are differences in
- Used Default Font (Windows -> Arial, Mac -> Helvetica Neue, Linux -> ?)
- Color Scheme (Same rendered pixel has different RGB values)

As Mac, Linux and Windows don't have a common default font, it's pretty hard to run the snapshot comparison on all environments with the same result, obviously. Also, on Mac there is the ColorSync which changes the RGB values of a rendered pixel.

Both leads to failing snapshot tests.

When increasing the `failureThreashold`, it starts missing real issues. E.g. changing the color of a curve in a chart will not be detected anymore.

Are there any configuration options known, maybe for Cypress or Chrome that will lead to similar rendering results on all systems?

Thanks!

**Comment by simonsmith** [comment]
Hey @jwedel, I moved it to a discussion

Have you tried Docker for this? I've had a similar issue on my projects (difference between Mac and Linux CI machines) and I generate my snapshots using run mode in Docker. I then rely on open mode for testing

**Reply by jwedel** [reply]
Thanks for the quick reply! 

Though about this too but was afraid of the effort a bit, TBH. How do you do this during development? 

- Do you start the app inside the docker container?
- Do you mount the app / tests into the container?
- Do you use a cypress base image or one that run the app (nginx etc)?

Would be great, if you could briefly describe the process that works for you!

**Reply by simonsmith** [reply]
So the key to making it simple is to mount the current working directory of your project inside the Docker container, so it's reading and writing to your project, not copying the app files into the container.

Here are two articles that might help:
* https://medium.com/testing-with-cypress/running-cypress-tests-in-docker-containers-using-different-docker-images-2dee3450881e
* https://www.cypress.io/blog/2019/05/02/run-cypress-with-a-single-docker-command

I think the key command is something like this:

```
docker run -it --env CYPRESS_updateSnapshots -v $PWD:/e2e -w /e2e cypress/included:12.8.1
```

The `-v` is mounting your current working directory into the container as `/e2e` and then running the Cypress command from there with `-w`

Some things to note that I do differently where I work:
* We have our own Docker image (mainly because we have dedicated ubuntu/node images) but the ones that Cypress provide are very good. We only use Electron because Chrome, Firefox etc don't work on ARM chips
* We have a basic CLI tool that will attempt to run different `yarn` commands inside the container, and these map back to scripts inside the `package.json` of the project. This let's each project have fine control. You might not need that level of complexity for now, but it for example allows us to have different `yarn` scripts for running component testing mode.
* The article from Cypress advocates running open mode in the container as well, which we tried but it's nicer and faster to run open mode normally, and just use `run` mode in the container. During open mode test failures from snapshots are ignored

**Reply by jwedel** [reply]
Thanks so much for the detailed comment! I will certainly try it!

**Reply by jwedel** [reply]
@simonsmith I just tried it and ran into one issue. How do you connect Cypress to the host? Also the linked article does not cover this. Do you start your app outside the container and than expose the host network or do you run the app inside the container?

**Reply by simonsmith** [reply]
Usually it's triggered from inside the container, so it might run a `yarn` script that starts a mock server and then runs the tests. Here's an example of what those scripts look like:

```json
{
    "cy:run": "cypress run",
    "test:e2e:run": "LOG_LEVEL=silent npm-run-all --parallel --race start:e2e start:e2e:server 'cy:run {@}' --",
    "start:e2e": "API_URI=localhost:3069 yarn start",
    "start:e2e:server": "ts-node cypress/server/e2e.ts"
}
```

Our CLI script uses the following function to run Docker:

```bash
docker_run() {
  exec \
    docker run -it --rm \
      --network host \
      -e CYPRESS_updateSnapshots=$CYPRESS_updateSnapshots \
      -v $PWD:/e2e \
      -w /e2e \
      --user 1000:1000 \
      --entrypoint="yarn" \
      $CYPRESS_IMAGE \
      $@
}
````

And then later:

```bash
docker_run test:e2e $@
```

So the `yarn` scripts are started in the container itself

Looks like the `--network` flag is what you need?

**Reply by jwedel** [reply]
For anyone interested:

We could not get it to run the app inside the container as we use `nx` and nx apparently uses platform-specific binaries. So if you do
- `yarn install` (Mac)
- `yarn run start:mocks` (Inside Linux container)

The latter fails. As we did not want to run a `yarn install` inside the container, we were looking for alternatives. We got it running by changing the baseUrl (`--config baseUrl=http://host.docker.internal:4200`) to connect to the host port 4200 where our application runs. 

So we start the app on the host and then use the cypress base image to run the tests inside the container.

The created snapshots now work pretty well also on CI where Linux is used.

So, big thanks again to @simonsmith for the valuable hints!

**Reply by simonsmith** [reply]
@jwedel Great to hear! Thanks for adding your findings too 👌🏻
